### PR TITLE
Focus skip link on navigation

### PR DESCRIPTION
**Pull request by madalynrose**
## Description

Many users have difficulty orienting themselves on a new page if focus is set on too large of an element or an inoperable element (like our wrapper `div`). Sending focus directly to a smaller, interactive control like a skip link is ideal. Unfortunately, we're limited with what we can programmatically achieve at the framework level (we have no way of knowing if a skip link exists on site pages from our `Router`). Focus on a skip link on navigation (as recommended based on @marcysutton's research with Fable Tech Labs in #20540) must be achieved on the site level. Since we'll recommend developers do this on their end, it makes sense to implement it on Gatsby's .org site as well.  We should also add an example of how to achieve this (#21107).

Because we're an organization and have to weigh the needs of all of our users, we'll have to weigh tradeoffs, specifically around the styling of the skip link. Default behavior both focuses and un-hides the skip link with every navigation, even when using a mouse. This is the recommended experience, since low-vision users who rely on magnification need this visual. It might be worth thinking about how to implement user preferences (e.g. saving in local storage) to toggle between showing on focus or keeping it visually hidden.

Feel free to test out this behavior [here](https://build-edbd21b4-6738-4682-b550-ea6b7176d248.gtsb.io/blog/).

### CODE REVIEW

**Comment by shannonbux** [comment]
Just a couple thoughts here:
- It sounds to me like "skip to content" basically behaves like the leftmost item on the main navigation even though it might look like it's not part of main navigation right now. If that's the case, I wonder if actually adding it as the leftmost item on the main navigation might work for these reasons:
- it would cause the least disruption for sighted users because it wouldn't block the view of any elements behind it. Right now, there's not a lot of white space in the upper lefthand corner.
- it would still provide the necessary functionality to low-vision users or those using a screen reader. It could look like a normal nav item except have a little fun icon that links to a page explaining why it's there, perhaps. 

The other question I have: if someone is navigating from one page to another via the keyboard, upon arrival at a new page would it ever cause problems for them to have the focus set to the "skip to content" button? In my mind, it would just be an advantage, even for sighted users. I was just testing this out and I think most people will appreciate being able to skip tabbing thru the nav. Just want to make sure I'm right about that.

**Comment by shannonbux** [comment]
A conversation after some hallway testing that @madalynrose has already seen, just realized I ought to include it here:

Shannon Soper Mar 27 2020, 9:04 am 
Hi @madalynrose! Here are the comments from @fk_ and @jeremybarnes. I showed them the PR in our design review meeting this week:
Requests:
this looks good as long as the skip to content button doesn’t hide visual navigation. One idea; it could appear to the left of the Gatsby logo.

Questions:

How about hitting tab first to trigger the behavior? Why does it need to show up right away? Is there a use case where someone with low vision would not use tab as their first action anyway?
What if we detect magnification? Or application (screen reader software)? Is it possible to detect these to trigger the skip link nav button to show up without the person needing to hit tab first? And then for everyone else, it's necessary to hit tab first before seeing it?
Different grading levels for accessibility. What grade do we shoot for?

Madalyn ParkerMadalyn Parker Mar 27 2020, 11:07 am (edited) 
HI there @shannonsoper! Great questions!

> How about hitting tab first to trigger the behavior? Why does it need to show up right away? Is there a use case where someone with low vision would not use tab as their first action anyway?

so a lot of people with low vision would be using a mouse still and have the page magnified. we want to focus their attention immediately on that link. this is from Marcy's research, specifically this section

> What if we detect magnification? Or application (screen reader software)? Is it possible to detect these to trigger the skip link nav button to show up without the person needing to hit tab first? And then for everyone else, it's necessary to hit tab first before seeing it?

detection like this is something that would be super useful for accessibility in theory but isn't possible right now, nor is it desirable for most people considering privacy and the probable case that many developers might use this detection to provide a lesser experience to those users.

> Different grading levels for accessibility. What grade do we shoot for?

could you expand more on this?

SS
Shannon Soper Mar 27 2020, 1:52 pm 
Thank you for the clarification, @madalynrose! Knowing that screen magnification users would likely use a mouse answers the first question.

To expand more on "different grading levels for accessibility. What grade do we shoot for?" I think @jeremybarnes would be best to provide more details; he mentioned that at a previous company they had different "grades" for how accessible their products were, to help with goal setting (aka, this is at grade level D, let's bring it to a C).

File: www/gatsby-browser.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+exports.onRouteUpdate = ({ location, prevLocation }) => {
```
Comment (pvdz) [inline www/gatsby-browser.js:1]: `location` is unused. I wonder why linting is not complaining about this...

**Comment by LekoArts** [comment]
This will soon be superseded by https://github.com/gatsbyjs/gatsby/pull/26376

**Pull request closed without merging**